### ISSUE TITLE
Google gRPC over TLS connection fails with INTERNAL error in Envoy 1.31.0

### ISSUE THREAD

**Issue by kubrickfr** [envoyproxy/envoy#35940]
Description:
When using google_grpc for ADS and load stats configuration in Envoy 1.31.0, the connection fails with an INTERNAL error (status code 13). The same configuration works correctly in Envoy 1.30.0.

Version:
- Envoy version with issue: 1.31.0
- Last known working version: 1.30.4

Configuration:
```yaml
node:
  id: test-envoy-node
  cluster: myclustername
  locality:
    region: "eu-west-1"
    zone: "eu-west-1b"
    sub_zone: "nil"

dynamic_resources:
  lds_config:
    resource_api_version: V3
    ads: {}
  cds_config:
    resource_api_version: V3
    ads: {}
  ads_config:
    transport_api_version: V3
    api_type: DELTA_GRPC
    grpc_services:
    - google_grpc:
        target_uri: my.control.plane.com:443
        stat_prefix: ads
        channel_credentials:
          ssl_credentials:
            root_certs:
              filename: /etc/ssl/certs/ca-certificates.crt
    set_node_on_first_message_only: true
    initial_metadata:
    - key: access_key
      value: xxx

cluster_manager: 
  load_stats_config: 
    transport_api_version: V3
    api_type: DELTA_GRPC 
    grpc_services: 
    - google_grpc: 
        target_uri: my.control.plane.com:443 
        stat_prefix: load_stats 
        channel_credentials:
          ssl_credentials:
            root_certs:
              filename: /etc/ssl/certs/ca-certificates.crt
    set_node_on_first_message_only: true
    initial_metadata:
    - key: access_key
      value: xxx

admin:
  address:
    socket_address:
      address: 0.0.0.0
      port_value: 9901
```

Error message:
```
[2024-09-02 15:31:55.952][1585][debug][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:54] Establishing new gRPC bidi stream to my.control.plane.com:443 for rpc DeltaAggregatedResources(stream .envoy.service.discovery.v3.DeltaDiscoveryRequest) returns (stream .envoy.service.discovery.v3.DeltaDiscoveryResponse);

[2024-09-02 15:31:55.952][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 0 false
[2024-09-02 15:31:55.952][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:268] Queued message to write (40755 bytes)
[2024-09-02 15:31:55.952][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=0 ok=false inflight=1
[2024-09-02 15:31:55.952][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:254] notifyRemoteClose 13 
[2024-09-02 15:31:55.952][1585][warning][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:176] DeltaAggregatedResources gRPC config stream to my.control.plane.com:443 closed: 13, 
[2024-09-02 15:31:55.952][1585][debug][config] [source/extensions/config_subscription/grpc/grpc_subscription_impl.cc:125] gRPC update for type.googleapis.com/envoy.config.cluster.v3.Cluster failed
[2024-09-02 15:31:55.952][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:283] resetStream
[2024-09-02 15:31:55.953][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:456] Stream cleanup with 1 in-flight tags
[2024-09-02 15:31:55.953][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 5 true
[2024-09-02 15:31:55.953][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=5 ok=true inflight=1
[2024-09-02 15:31:55.953][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:445] Deferred delete
[2024-09-02 15:31:55.953][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:194] GoogleAsyncStreamImpl destruct
[2024-09-02 15:31:56.196][1585][debug][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:54] Establishing new gRPC bidi stream to my.control.plane.com:443 for rpc DeltaAggregatedResources(stream .envoy.service.discovery.v3.DeltaDiscoveryRequest) returns (stream .envoy.service.discovery.v3.DeltaDiscoveryResponse);

[2024-09-02 15:31:56.196][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 0 false
[2024-09-02 15:31:56.196][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:268] Queued message to write (40755 bytes)
[2024-09-02 15:31:56.196][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=0 ok=false inflight=1
[2024-09-02 15:31:56.196][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:254] notifyRemoteClose 13 
[2024-09-02 15:31:56.196][1585][warning][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:176] DeltaAggregatedResources gRPC config stream to my.control.plane.com:443 closed: 13, 
[2024-09-02 15:31:56.196][1585][debug][config] [source/extensions/config_subscription/grpc/grpc_subscription_impl.cc:125] gRPC update for type.googleapis.com/envoy.config.cluster.v3.Cluster failed
[2024-09-02 15:31:56.196][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:283] resetStream
[2024-09-02 15:31:56.196][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:456] Stream cleanup with 1 in-flight tags
[2024-09-02 15:31:56.196][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 5 true
[2024-09-02 15:31:56.196][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=5 ok=true inflight=1
[2024-09-02 15:31:56.196][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:445] Deferred delete
[2024-09-02 15:31:56.196][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:194] GoogleAsyncStreamImpl destruct
[2024-09-02 15:31:57.325][1585][debug][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:54] Establishing new gRPC bidi stream to my.control.plane.com:443 for rpc DeltaAggregatedResources(stream .envoy.service.discovery.v3.DeltaDiscoveryRequest) returns (stream .envoy.service.discovery.v3.DeltaDiscoveryResponse);

[2024-09-02 15:31:57.325][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 0 false
[2024-09-02 15:31:57.325][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:268] Queued message to write (40755 bytes)
[2024-09-02 15:31:57.325][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=0 ok=false inflight=1
[2024-09-02 15:31:57.325][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:254] notifyRemoteClose 13 
[2024-09-02 15:31:57.325][1585][warning][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:176] DeltaAggregatedResources gRPC config stream to my.control.plane.com:443 closed: 13, 
[2024-09-02 15:31:57.325][1585][debug][config] [source/extensions/config_subscription/grpc/grpc_subscription_impl.cc:125] gRPC update for type.googleapis.com/envoy.config.cluster.v3.Cluster failed
[2024-09-02 15:31:57.325][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:283] resetStream
[2024-09-02 15:31:57.325][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:456] Stream cleanup with 1 in-flight tags
[2024-09-02 15:31:57.325][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 5 true
[2024-09-02 15:31:57.325][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=5 ok=true inflight=1
[2024-09-02 15:31:57.325][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:445] Deferred delete
[2024-09-02 15:31:57.325][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:194] GoogleAsyncStreamImpl destruct
[2024-09-02 15:31:57.433][1585][debug][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:54] Establishing new gRPC bidi stream to my.control.plane.com:443 for rpc DeltaAggregatedResources(stream .envoy.service.discovery.v3.DeltaDiscoveryRequest) returns (stream .envoy.service.discovery.v3.DeltaDiscoveryResponse);

[2024-09-02 15:31:57.433][1589][trace][grpc] [source/common/grpc/google_async_client_impl.cc:61] completionThread CQ event 0 false
[2024-09-02 15:31:57.433][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:268] Queued message to write (40755 bytes)
[2024-09-02 15:31:57.433][1585][trace][grpc] [source/common/grpc/google_async_client_impl.cc:339] handleOpCompletion op=0 ok=false inflight=1
[2024-09-02 15:31:57.433][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:254] notifyRemoteClose 13 
[2024-09-02 15:31:57.433][1585][warning][config] [./source/extensions/config_subscription/grpc/grpc_stream.h:176] DeltaAggregatedResources gRPC config stream to my.control.plane.com:443 closed: 13, 
[2024-09-02 15:31:57.433][1585][debug][config] [source/extensions/config_subscription/grpc/grpc_subscription_impl.cc:125] gRPC update for type.googleapis.com/envoy.config.cluster.v3.Cluster failed
[2024-09-02 15:31:57.433][1585][debug][grpc] [source/common/grpc/google_async_client_impl.cc:283] resetStream

```

Steps to reproduce:

Set up a control plane server that accepts gRPC connections over TLS.
Configure Envoy 1.31.0 with the provided configuration.
Start Envoy and observe the logs.

Expected behavior:
Envoy should establish a gRPC connection to the control plane successfully.
Actual behavior:
Envoy fails to establish the gRPC connection, reporting an INTERNAL error (status code 13).
Additional information:

The same configuration works correctly with Envoy 1.30.4.
The control plane server is reachable and properly configured (verified using curl and working with Envoy 1.30.4).
No firewall or network issues are present (verified by successful connections using curl and Envoy 1.30.4).

Possible related changes:
Are there any changes in the gRPC implementation or TLS handling between Envoy 1.30.0 and 1.31.0 that could explain this behavior?
Any assistance in identifying and resolving this issue would be greatly appreciated.

**Comment by kubrickfr** [comment]
NB, I wrote "over TLS" but what I mean is "Over HTTP2 with TLS", if it makes a difference... The endpoint is an AWS ALB.

**Comment by phlax** [comment]
cc @adisuissa

**Comment by adisuissa** [comment]
I believe this was addressed by #35487 and backported in #35511.
If you can use any of these PRs it should fix the problem.
I think v1.31.1 should be out in mid-October.

**Comment by phlax** [comment]
1.31.1 is due on 19th September - see https://github.com/envoyproxy/envoy/pull/35909

**Comment by adisuissa** [comment]
Oh, good to know. I see that it is the security release. Will it also include the backports?

**Comment by phlax** [comment]
yep

**Comment by phlax** [comment]
there was some discussion about cutting a backport release in advance - but no definite plan for that cc @ggreenway

**Comment by kubrickfr** [comment]
> I believe this was addressed by #35487 and backported in #35511. If you can use any of these PRs it should fix the problem. I think v1.31.1 should be out in mid-October.

I confirm that #35511 fixes the problem, thank you very much!
